### ISSUE TITLE
Protected field namings not checked when 1304 and 1307 enabled simultaneously

### ISSUE THREAD

**Issue by tobias-wiese** [DotNetAnalyzers/StyleCopAnalyzers#3557]
Consider the following mininal example for C#:
```cs
namespace Some.Space
{
    public abstract class SomeClass
    {
        protected readonly string foo;
        internal readonly string bar;
    }
}
```

I have a test configuration with only rules SA1304 (“The name of a non-private readonly C# field should being with an upper-case letter.”) and SA1307 (“The name of a public or internal field in C# does not begin with an upper-case letter.”) enabled.

By this means, we would expect SA1304 firing for `foo` and `bar`, and SA1307 firing for `bar` . 
However, the output shows two warnings of these rules, both only for `bar` .

I think this is caused by this code snippet in https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1304NonPrivateReadonlyFieldsMustBeginWithUpperCaseLetter.cs#L82
```cs
if (!syntax.Modifiers.Any(SyntaxKind.InternalKeyword))
{
    // SA1307 is taken precedence here. SA1307 should be reported if the field is accessible.
    // So if SA1307 is enabled this diagnostic will only be reported for internal fields.
    if (!context.IsAnalyzerSuppressed(SA1307AccessibleFieldsMustBeginWithUpperCaseLetter.Descriptor))
    {
        return;
    }
}
```

This basically skips the SA1304 check if the field keyword is `public` or `protected` (i.e. not `internal`) and SA1307 is enabled. 

However, SA1307 does not consider `protected` fields, see the code in https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/StyleCop.Analyzers/StyleCop.Analyzers/NamingRules/SA1307AccessibleFieldsMustBeginWithUpperCaseLetter.cs#L74
```cs
if (declaration.Modifiers.Any(SyntaxKind.PublicKeyword) || declaration.Modifiers.Any(SyntaxKind.InternalKeyword))
```

 This way, `protected` fields are never checked if SA1304 and SA1307 are enabled simultaneously, even if they should?

**Comment by bjornhellander** [comment]
This has probaly been around since 2015. Looks like a mistake, but there are actually tests that verify this behaviour. Unclear why though, since the commit has no info on that.

Besides the cases you mention, there are also some "newer" keyword combinations to consider. And I don't understand why SA1304 is not skipped for internal.

This code shows what happens now (and what I think should happen, where it differs, based on the documentation of SA1304 and SA1307):
```
public class TestClass
{
    public readonly int a1; // SA1307
    public int a2; // SA1307

    protected internal readonly int c1; // SA1304 and SA1307. Should only trigger SA1307.
    protected internal int c2; // SA1307

    internal readonly int b1; // SA1304 and SA1307. Should only trigger SA1307.
    internal int b2; // SA1307

    protected readonly int d1; // None. Should trigger SA1304.
    protected int d2; // None.

    private protected readonly int e1; // None. Should trigger SA1304.
    private protected int e2; // None

    private readonly int f1; // None
    private int f2; // None
}
```

What do you think, @sharwell?

**Comment by tobias-wiese** [comment]
@sharwell @bjornhellander Any news on this? :) In other words: can I expect this issue to be fixed, or just work with the checks as they are?
